internal/pkg/upload/upload.go

```diff
@@ -0,0 +1,286 @@
+	AgentID   string
+	ActionID  string
+	ChunkSize int64
+	Total     int64
+	Count     int
+	Start     time.Time
+	Status    Status
+}
+
+// convenience functions for computing current "Status" based on the fields
+func (i Info) Expired(timeout time.Duration) bool { return time.Now().After(i.Start.Add(timeout)) }
+func (i Info) StatusCanUpload() bool { // returns true if more chunks can be uploaded. False if the upload process has completed (with or without error)
+	return !(i.Status == StatusFail || i.Status == StatusDone || i.Status == StatusDel)
+}
+
+type FileData struct {
+	Size      int64  `json:"size"`
+	ChunkSize int64  `json:"ChunkSize"`
+	Status    string `json:"Status"`
+}
+
+type FileMetaDoc struct {
+	ActionID string    `json:"action_id"`
+	AgentID  string    `json:"agent_id"`
+	Source   string    `json:"src"`
+	File     FileData  `json:"file"`
+	UploadID string    `json:"upload_id"`
+	Start    time.Time `json:"upload_start"`
+}
+
+// custom unmarshaller to make unix-epoch values work
+func (f *FileMetaDoc) UnmarshalJSON(b []byte) error {
+	type InnerFile FileMetaDoc // type alias to prevent recursion into this func
+	// override the field to parse as an int, then manually convert to time.time
+	var tmp struct {
+		InnerFile
+		Start int64 `json:"upload_start"`
+	}
+	if err := json.Unmarshal(b, &tmp); err != nil {
+		return err
+	}
+
+	*f = FileMetaDoc(tmp.InnerFile) // copy over all fields
+	f.Start = time.UnixMilli(tmp.Start)
+	return nil
+}
+
+type ChunkInfo struct {
+	Pos  int  // Ordered chunk position in file
+	Last bool // Is this the final chunk in the file
+	SHA2 string
+	Size int
+	BID  string // base id, matches metadata doc's _id
+	//FirstReceived bool
+}
+
+type Uploader struct {
+	metaCache map[string]Info // cache of file metadata doc info
+	mu        sync.RWMutex    // lock for the above
+	// @todo: cache eviction so it's not unbounded growth
+	// @todo: cache refresh so status is accurate
+	sizeLimit int64         // @todo: what if configuration changes? is this recreated with another New()?
+	timeLimit time.Duration // @todo: same as above
+
+	chunkClient *elasticsearch.Client
+	bulker      bulk.Bulk
+}
+
+func New(chunkClient *elasticsearch.Client, bulker bulk.Bulk, sizeLimit int64, timeLimit time.Duration) *Uploader {
+	return &Uploader{
+		chunkClient: chunkClient,
+		bulker:      bulker,
+		sizeLimit:   sizeLimit,
+		timeLimit:   timeLimit,
+		metaCache:   make(map[string]Info),
+	}
+}
+
+// Start an upload operation
+func (u *Uploader) Begin(ctx context.Context, data JSDict) (Info, error) {
+	if data == nil {
+		return Info{}, errors.New("upload start payload required")
+	}
+
+	/*
+		Validation and Input parsing
+	*/
+
+	// make sure all required fields are present and non-empty
+	if err := validateUploadPayload(data); err != nil {
+		return Info{}, err
+	}
+
+	size, _ := data.Int64("file", "size")
+	if size > u.sizeLimit {
+		return Info{}, ErrFileSizeTooLarge
+	}
```

Comment:
need to check if size is <= 0 ?

Reply:
`<= 0` would be an input error ( illogical for negative file size, useless upload of 0 size) so these cases are filtered out during input validation in `upload.go:260`, `func validateUploadPayload`.

Anything past that is valid input (i.e. a positive non-zero file size). And we check against server limits later (here)